kms/awskms/signer.go

```diff
@@ -0,0 +1,122 @@
+package awskms
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"io"
+
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/pkg/errors"
+	"github.com/smallstep/cli/crypto/pemutil"
+)
+
+// Signer implements a crypto.Signer using the AWS KMS.
+type Signer struct {
+	service   KeyManagementClient
+	keyID     string
+	publicKey crypto.PublicKey
+}
+
+// NewSigner creates a new signer using a key in the AWS KMS.
+func NewSigner(svc KeyManagementClient, signingKey string) (*Signer, error) {
+	keyID, err := parseKeyID(signingKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// Make sure that the key exists.
+	signer := &Signer{
+		service: svc,
+		keyID:   keyID,
+	}
+	if err := signer.preloadKey(keyID); err != nil {
```

Comment:
There's no standard around that, I basically did it to detect an error earlier in our code rather than when the key is going to be used inside the `crypto/x509` package at the signing time, way later than the signer is created.

Reply:
Normally I would expect creating a New thing would just give me the struct initialized as needed then if I wanted to preload the key I could do that myself (possibly in a goroutine) right after making a new one. Something like:

```
s, err := NewSigner(svc, kid)
if err != nil {...}
err = s.PreloadKey() or go s.PreloadKey()
```
This just puts more control in the hands of the caller as to when network/long-running tasks are being performed.

Reply:
But this is easy enough to change if needed...

Reply:
You can, but you're not supposed to use the NewSigner directly, only if you just want to support awskms you will do that. So the way we use it we'll have to call signer.Public() to detect the error, and do not wait until somebody signs a key, that can be ... days later.